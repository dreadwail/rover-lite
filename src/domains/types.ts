import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
