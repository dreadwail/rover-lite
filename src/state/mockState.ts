import { RoverSearchState } from './types';

type Primitive = boolean | number | string;

export type RecursivePartial<ThisLevel> = {
  [Key in keyof ThisLevel]?: ThisLevel[Key] extends Primitive
    ? ThisLevel[Key]
    : RecursivePartial<ThisLevel[Key]>
};

export const createMockState = (
  interestingStateKeys?: RecursivePartial<RoverSearchState>
): RoverSearchState => (interestingStateKeys as unknown) as RoverSearchState;
