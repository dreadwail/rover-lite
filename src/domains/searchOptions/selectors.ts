import { Options, OptionsByCategoryId, OptionCategory } from '../../state/searchOptions/types';
import { RoverSearchState } from '../../state/types';

export const hasFailed = (state: RoverSearchState): boolean => state.searchOptions.hasFailed;
export const isFetching = (state: RoverSearchState): boolean => state.searchOptions.isFetching;

const getOptions = (state: RoverSearchState): Options => state.searchOptions.options;

export const getRankedCategories = (state: RoverSearchState): OptionCategory[] => {
  const options = getOptions(state);
  const categoriesByKey = Object.keys(options).reduce<Record<string, OptionCategory>>(
    (categories, key) => {
      const { category } = options[key];
      return { ...categories, [category.categoryId]: category };
    },
    {}
  );
  return Object.values(categoriesByKey).sort((a, b) => a.rank - b.rank);
};

export const getOptionsByCategoryId = (state: RoverSearchState): OptionsByCategoryId => {
  const options = getOptions(state);
  return Object.keys(options).reduce<OptionsByCategoryId>((byCategoryId, key) => {
    const option = options[key];
    const { category } = option;
    const optionsHere = byCategoryId[category.categoryId];
    const newOptionsHere = { ...optionsHere, [key]: option };
    return {
      ...byCategoryId,
      [category.categoryId]: newOptionsHere,
    };
  }, {});
};
