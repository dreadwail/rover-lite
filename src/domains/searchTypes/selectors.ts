import { SearchType } from '../../state/searchTypes/types';
import { RoverSearchState } from '../../state/types';

export const getSearchTypes = (state: RoverSearchState): SearchType[] => {
  const endpoints = state.searchTypes.endpoints;
  return Object.keys(endpoints) as SearchType[];
};
export const hasFailed = (state: RoverSearchState): boolean => state.searchTypes.hasFailed;
export const isFetching = (state: RoverSearchState): boolean => state.searchTypes.isFetching;
export const getEndpointUrl = (
  state: RoverSearchState,
  searchType: SearchType
): string | undefined => state.searchTypes.endpoints[searchType];
export const isSearchTypesLoaded = (state: RoverSearchState): boolean =>
  getSearchTypes(state).length > 0;
