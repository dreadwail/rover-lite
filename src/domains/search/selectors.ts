import { SearchInputs } from '../../state/search/types';
import { SearchType } from '../../state/searchTypes/types';
import { RoverSearchState } from '../../state/types';

export const getSearchType = (state: RoverSearchState): SearchType => state.search.searchType;
export const getInputs = (state: RoverSearchState): SearchInputs => state.search.inputs;
export const getLatLong = (state: RoverSearchState): [number, number] => [
  state.search.latitude,
  state.search.longitude,
];
