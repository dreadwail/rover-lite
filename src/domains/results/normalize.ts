import { Listing } from '../../state/results/types';

export interface RawDerivedData {
  readonly zip_code: string;
}

export interface RawResponsiveness {
  readonly photo_rate?: number;
  readonly photo_rate_desc: string;
  readonly percent?: number;
  readonly percent_desc: string;
  readonly time_short_desc: string;
}

export interface RawImage {
  readonly small: string;
}

export interface RawAttribute {
  readonly text: string;
  readonly key: string;
  readonly icon: string;
}

export interface RawPreferences {
  readonly dogs_preferences_attributes: RawAttribute[];
  readonly accepted_dogs_attributes: RawAttribute[];
  readonly dogs_experience_attributes: RawAttribute[];
  readonly attributes: RawAttribute[];
}

export interface RawProviderProfile {
  readonly hosting_service_preferences: RawPreferences;
  readonly traveling_service_preferences: RawPreferences;
  readonly attributes: RawAttribute[];
  readonly years_of_experience: number;
}

export interface RawBadge {
  readonly icon: string;
  readonly description: string;
  readonly title: string;
}

export interface RawPersonSummary {
  readonly responsiveness: RawResponsiveness;
  readonly first_name: string;
  readonly reviews_count: number;
  readonly profile_photo: RawImage;
  readonly ratings_average: string;
  readonly web_url: string;
  readonly images: RawImage[];
  readonly provider_profile: RawProviderProfile;
  readonly attributes: Record<string, RawAttribute>;
  readonly badges: RawBadge[];
}

export interface RawBadgeData {
  readonly content: string;
  readonly image_url: string;
  readonly show_on_search: boolean;
  readonly title: string;
}

export interface RawReview {
  readonly content: string;
  readonly dog_photo_url: string;
}

export interface RawResult {
  readonly rank: number;
  readonly person_summary: RawPersonSummary;
  readonly reviews_count: number;
  readonly neighborhood: string;
  readonly price_unit: string;
  readonly city: string;
  readonly zip: string;
  readonly title: string;
  readonly state: string;
  readonly contact_url: string;
  readonly person_name: string;
  readonly price: string;
  readonly review_text: string;
  readonly review_dog_photo_url: string;
  readonly badge_data: RawBadgeData[];
  readonly repeat_client_count: number;
  readonly reviews: RawReview[];
  readonly ratings_average: string;
  readonly currency_symbol: string;
}

export interface RawSearchResponse {
  readonly results: RawResult[];
}

const rawResultToListing = (rawResult: RawResult): Listing => ({
  thumbnailUrl: rawResult.person_summary.profile_photo.small,
  rank: rawResult.rank,
  name: rawResult.person_summary.first_name,
  badges: rawResult.badge_data.map(rawBadge => ({
    content: rawBadge.content,
    imageUrl: rawBadge.image_url,
    title: rawBadge.title,
  })),
  tagline: rawResult.title,
  neighborhood: rawResult.neighborhood,
  city: rawResult.city,
  state: rawResult.state,
  zip: rawResult.zip,
  repeatClientCount: rawResult.repeat_client_count,
  reviewCount: rawResult.reviews_count,
  ratingAverage: parseFloat(rawResult.ratings_average),
  featuredReview: rawResult.review_text,
  currencySymbol: rawResult.currency_symbol,
  price: parseFloat(rawResult.price),
  priceUnit: rawResult.price_unit,
  webUrl: rawResult.person_summary.web_url,
});

export const normalize = (rawResponse: RawSearchResponse): Listing[] =>
  rawResponse.results.map(rawResultToListing);
