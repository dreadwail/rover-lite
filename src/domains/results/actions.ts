import { Action } from 'redux';

import { request } from '../../api/http';
import { urlEncodeQuery } from '../../api/query';
import { Listing } from '../../state/results/types';
import { SearchInputs } from '../../state/search/types';
import debouncedAction from '../debouncedThunk';
import { getInputs, getLatLong, getSearchType } from '../search/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, RawSearchResponse } from './normalize';

export interface FetchResultsPendingAction extends Action {
  readonly type: 'FETCH_RESULTS.PENDING';
}

export interface FetchResultsFulfilledAction extends Action {
  readonly type: 'FETCH_RESULTS.FULFILLED';
  readonly payload: Listing[];
}

export interface FetchResultsRejectedAction extends Action {
  readonly type: 'FETCH_RESULTS.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchResultsPending = (): FetchResultsPendingAction => ({
  type: 'FETCH_RESULTS.PENDING',
});

const fetchResultsFulfilled = (listings: Listing[]): FetchResultsFulfilledAction => ({
  type: 'FETCH_RESULTS.FULFILLED',
  payload: listings,
});

const fetchResultsRejected = (error: any): FetchResultsRejectedAction => ({
  type: 'FETCH_RESULTS.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchResults = () => PromiseThunkAction<void>;
export const fetchResults: FetchResults = () => async (dispatch, getState) => {
  dispatch(fetchResultsPending());
  try {
    const state = getState();

    const searchType = getSearchType(state);
    const [latitude, longitude] = getLatLong(state);

    const defaultInputs: SearchInputs = {
      centerlat: latitude,
      centerlng: longitude,
      service_type: searchType,
    };
    const userInputs = getInputs(state);
    const inputs = { ...defaultInputs, ...userInputs };
    const encodedQuery = urlEncodeQuery(inputs);

    const urlBase = 'https://www.rover.com/api/v3/search';
    const url = `${urlBase}/${searchType}/?${encodedQuery}`;

    const response = await request<RawSearchResponse>('GET', url);

    const normalizedListings = normalize(response);
    dispatch(fetchResultsFulfilled(normalizedListings));
  } catch (err) {
    dispatch(fetchResultsRejected(err));
  }
};

export const debouncedFetchResults = debouncedAction(fetchResults, 250);
