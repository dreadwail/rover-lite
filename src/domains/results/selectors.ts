import { Listing } from '../../state/results/types';
import { RoverSearchState } from '../../state/types';

export const isFetching = (state: RoverSearchState): boolean => state.results.isFetching;
export const hasFailed = (state: RoverSearchState): boolean => state.results.hasFailed;
export const isStale = (state: RoverSearchState): boolean => state.results.isStale;
export const getListings = (state: RoverSearchState): Listing[] => state.results.listings;
