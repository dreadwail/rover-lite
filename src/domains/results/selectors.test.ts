import { createMockState } from '../../state/mockState';
import { Listing, Results } from '../../state/results/types';

import { getListings, hasFailed, isFetching, isStale } from './selectors';

describe('results selectors', () => {
  describe('.isFetching', () => {
    let selected: boolean;

    describe('when fetching', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: true,
          hasFailed: false,
          isStale: false,
          listings: [],
        };
        const state = createMockState({ results });
        selected = isFetching(state);
      });

      it('returns true', () => {
        expect(selected).toBe(true);
      });
    });

    describe('when not fetching', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: false,
          hasFailed: false,
          isStale: false,
          listings: [],
        };
        const state = createMockState({ results });
        selected = isFetching(state);
      });

      it('returns false', () => {
        expect(selected).toBe(false);
      });
    });
  });

  describe('.hasFailed', () => {
    let selected: boolean;

    describe('when fetching has failed', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: false,
          hasFailed: true,
          isStale: false,
          listings: [],
        };
        const state = createMockState({ results });
        selected = hasFailed(state);
      });

      it('returns true', () => {
        expect(selected).toBe(true);
      });
    });

    describe('when fetching hasnt failed', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: false,
          hasFailed: false,
          isStale: false,
          listings: [],
        };
        const state = createMockState({ results });
        selected = hasFailed(state);
      });

      it('returns false', () => {
        expect(selected).toBe(false);
      });
    });
  });

  describe('.isStale', () => {
    let selected: boolean;

    describe('when results are stale', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: false,
          hasFailed: false,
          isStale: true,
          listings: [],
        };
        const state = createMockState({ results });
        selected = isStale(state);
      });

      it('returns true', () => {
        expect(selected).toBe(true);
      });
    });

    describe('when results are fresh', () => {
      beforeEach(() => {
        const results: Results = {
          isFetching: false,
          hasFailed: false,
          isStale: false,
          listings: [],
        };
        const state = createMockState({ results });
        selected = isStale(state);
      });

      it('returns false', () => {
        expect(selected).toBe(false);
      });
    });
  });

  describe('.getListings', () => {
    let selected: Listing[];

    beforeEach(() => {
      const results: Results = {
        isFetching: false,
        hasFailed: false,
        isStale: false,
        listings: [
          {
            thumbnailUrl: 'a',
            rank: 0,
            name: 'b',
            badges: [],
            tagline: 'c',
            neighborhood: 'Ballard',
            city: 'Seattle',
            state: 'WA',
            zip: '98117',
            repeatClientCount: 3,
            reviewCount: 4,
            ratingAverage: 4.5,
            featuredReview: 'd',
            currencySymbol: '$',
            price: 42.0,
            priceUnit: 'night',
            webUrl: 'e',
          },
        ],
      };
      const state = createMockState({ results });
      selected = getListings(state);
    });

    it('returns the current listings', () => {
      expect(selected).toMatchSnapshot();
    });
  });
});
