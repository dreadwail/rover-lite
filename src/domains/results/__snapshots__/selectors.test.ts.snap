// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results selectors .getListings returns the current listings 1`] = `
Array [
  Object {
    "badges": Array [],
    "city": "Seattle",
    "currencySymbol": "$",
    "featuredReview": "d",
    "name": "b",
    "neighborhood": "Ballard",
    "price": 42,
    "priceUnit": "night",
    "rank": 0,
    "ratingAverage": 4.5,
    "repeatClientCount": 3,
    "reviewCount": 4,
    "state": "WA",
    "tagline": "c",
    "thumbnailUrl": "a",
    "webUrl": "e",
    "zip": "98117",
  },
]
`;
